### PR TITLE
Add interrupt handler to gpbackup and gprestore

### PR DESCRIPTION
**Pull request by jmcatamney**
To ensure that the cluster is not left in a bad state if someone aborts a backup
or restore with Ctrl-C, this commit moves cleanup-related functions that were
previously called with defer to a DoCleanup function and sets up a signal handler
goroutine to call DoCleanup and exit if an interrupt is received.  DoTeardown
and several other functions are modified to account for a possible abort in the
middle of the function.

Author: Jamie McAtamney <jmcatamney@pivotal.io>

### CODE REVIEW

File: utils/db.go [hunk 1]
```diff
@@ -278,3 +278,15 @@ func SelectStringSlice(connection *DBConn, query string, whichConn ...int) []str
 	}
 	return retval
 }
+
+// TODO: Uniquely identify COPY commands in the multiple data file case to allow terminating sessions
```
Comment (chrishajas) [inline utils/db.go]: Does this TODO still apply? It seems like we're identifying the COPY by its PID.

File: restore/remote.go [hunk 2]
```diff
@@ -72,7 +72,7 @@ func CleanUpHelperFilesOnAllHosts(cluster utils.Cluster) {
 	remoteOutput := cluster.GenerateAndExecuteCommand("Removing oid list and helper script files from segment data directories", func(contentID int) string {
 		oidFile := cluster.GetSegmentHelperFilePath(contentID, "oid")
 		scriptFile := cluster.GetSegmentHelperFilePath(contentID, "script")
-		return fmt.Sprintf("rm %s && rm %s", oidFile, scriptFile)
+		return fmt.Sprintf("rm -f %s && rm -f %s", oidFile, scriptFile)
```
Comment (chrishajas) [inline restore/remote.go]: Is there a reason for `-f` here? Is it possible for these files to not exist/be cleaned up previously?

File: restore/wrappers.go [hunk 3]
```diff
@@ -30,11 +32,30 @@ func SetLoggerVerbosity() {
...
+SET gp_default_storage_options='';
+`
+	for i := 0; i < connection.NumConns; i++ {
+		connection.MustExec(setupQuery, i)
+	}
+}
+
+func InitializeSignalHandler() {
+	signalChan := make(chan os.Signal, 1)
+	signal.Notify(signalChan, os.Interrupt)
```
Comment (chrishajas) [inline restore/wrappers.go]: Do we handle a `kill` signal here in addition to interrupt?

File: restore/remote.go [hunk 4]
```diff
@@ -141,10 +141,36 @@ func VerifyHelperVersionOnSegments(cluster utils.Cluster, version string) {
...
+	cluster.CheckClusterError(remoteOutput, "Unable to clean up segment data pipes", func(contentID int) string {
+		return "Unable to clean up segment data pipe"
+	})
+}
+
 func CleanUpSegmentTOCs(cluster utils.Cluster) {
 	remoteOutput := cluster.GenerateAndExecuteCommand("Removing segment table of contents files from segment data directories", func(contentID int) string {
 		tocFile := cluster.GetSegmentTOCFilePathWithPID(cluster.SegDirMap[contentID], fmt.Sprintf("%d", contentID))
-		return fmt.Sprintf("rm %s", tocFile)
+		return fmt.Sprintf("rm -f %s", tocFile)
```
Comment (chrishajas) [inline restore/remote.go]: Is the `-f` necessary? In wht case would the TOC file not exist?

File: restore/remote.go [hunk 5]
```diff
@@ -141,10 +141,36 @@ func VerifyHelperVersionOnSegments(cluster utils.Cluster, version string) {
...
+		 * the time DoCleanup is called.
+		 */
+		return fmt.Sprintf("PIDS=`ps ux | grep \"%s\" | grep \"%s\" | grep -v grep | awk '{print $2}'`; if [[ ! -z \"$PIDS\" ]]; then kill -9 $PIDS; fi", procPattern)
+	})
+	cluster.CheckClusterError(remoteOutput, "Unable to clean up restore agent processes", func(contentID int) string {
+		return "Unable to clean up restore agent process"
+	})
+}
+
+func CleanUpSegmentPipesOnAllHosts(cluster utils.Cluster) {
```
Comment (chrishajas) [inline restore/remote.go]: I see this is used on the backup side, but is this function used anywhere for restore?

Comment (khuddlefish) [reply restore/remote.go]: It does not appear to be and shouldn't be. It won't actually do the right thing if called on the restore side.

File: restore/restore.go [hunk 6]
```diff
@@ -225,12 +237,41 @@ func DoTeardown() {
...
+	defer func() {
+		if err := recover(); err != nil {
+			logger.Warn("Encountered error during cleanup: %v", err)
+		}
+		logger.Verbose("Cleanup complete")
+		CleanupGroup.Done()
+	}()
+	logger.Verbose("Beginning cleanup")
+	if backupConfig.SingleDataFile {
+		CleanUpSegmentHelperProcesses(globalCluster)
```
Comment (chrishajas) [inline restore/restore.go]: In a case with no error (eg: normal gprestore exit), the helper processes should have already exited and not required cleanup right?

File: restore/remote.go [hunk 7]
```diff
@@ -141,10 +141,36 @@ func VerifyHelperVersionOnSegments(cluster utils.Cluster, version string) {
...
+	remoteOutput := cluster.GenerateAndExecuteCommand("Cleaning up segment data pipes", func(contentID int) string {
+		pipePath := cluster.GetSegmentPipeFilePath(contentID)
```
Comment (khuddlefish) [inline restore/remote.go]: If this was used on the restore side, this should be GetSegmentPipeFilePathWithPID. It might actually make sense for us to move those functions to backup/restore respectively. Maybe not part of this commit though.

File: end_to_end/end_to_end_suite_test.go [hunk 8]
```diff
@@ -318,5 +330,54 @@ var _ = Describe("backup end to end integration tests", func() {
...
+				rng := rand.New(rand.NewSource(time.Now().UnixNano()))
+				time.Sleep(time.Duration(rng.Intn(1000)+500) * time.Millisecond)
+				cmd.Process.Signal(os.Interrupt)
+			}()
+			output, _ := cmd.CombinedOutput()
+			stdout := string(output)
+
+			Expect(stdout).To(ContainSubstring("Received an interrupt, aborting backup process"))
+			Expect(stdout).To(ContainSubstring("Cleanup complete"))
+			Expect(stdout).To(Not(ContainSubstring("CRITICAL")))
```
Comment (chrishajas) [inline end_to_end/end_to_end_suite_test.go]: Would it be worth doing a `ps ux` and asserting no processes still exist anywhere, or is that a bit overkill?

File: backup/backup.go [hunk 9]
```diff
@@ -262,8 +274,33 @@ func DoTeardown() {
...
+			logger.Warn("Encountered error during cleanup: %v", err)
+		}
+		logger.Verbose("Cleanup complete")
+		CleanupGroup.Done()
+	}()
+	logger.Verbose("Beginning cleanup")
+	if *singleDataFile {
+		CleanUpSegmentPipesOnAllHosts(globalCluster)
+		CleanUpSegmentTailProcesses(globalCluster)
+		utils.TerminateHangingCopySessions(connection, globalCluster, "gpbackup")
```
Comment (chrishajas) [inline backup/backup.go]: We also don't necessarily have to run this if we exit normally.

File: backup/backup.go [hunk 10]
```diff
@@ -230,12 +234,20 @@ func DoTeardown() {
...
+		/*
+		 * Don't print an error or create a report file if the backup was canceled,
+		 * as the signal handler will take care of cleanup and return codes.  Just
+		 * wait until the signal handler's DoCleanup completes so the main goroutine
+		 * doesn't exit while cleanup is still in progress.
+		 */
+		CleanupGroup.Wait()
+		return
 	}
+	fmt.Println(errStr)
```
Comment (chrishajas) [inline backup/backup.go]: When errStr is empty, it will still print an empty line. Is this desired?

File: utils/db.go (hunk 1 above)
Comment (jmcatamney) [reply utils/db.go]: We identify the COPY by its PID in the single file case only, not in the multiple-file case (c.f. GetTableBackupFilePathForCopyCommand).

File: restore/remote.go (hunk 2 above)
Comment (jmcatamney) [reply restore/remote.go]: Yes, in two cases.  If a Ctrl-C is received during DoTeardown both the normal DoCleanup and the signal handler DoCleanup will try to remove them, and if restore errors out early (e.g. trying to --include-schema a bad schema) it'll still call DoCleanup even if no files were created yet.

File: restore/wrappers.go (hunk 3 above)
Comment (jmcatamney) [reply restore/wrappers.go]: Just interrupt for now, as that's all that was in the story and we may leave to allow a `kill -9` for "No, seriously, die now and don't try to clean up" cases.

File: restore/remote.go (hunk 4 above)
Comment (jmcatamney) [reply restore/remote.go]: If the cleanup happens before the files are created, as mentioned in a previous comment, we don't want a bunch of extraneous error messages at the end.

File: restore/remote.go (hunk 5 above)
Comment (jmcatamney) [reply restore/remote.go]: Yep, just copy-pasted it into restore/remote.go by mistake when I was moving some things around.  I'll delete it.

Comment (chrishajas) [reply restore/remote.go]: I think there's already an identical function in backup/remote.go actually, so I think this function can just be removed.

File: restore/restore.go (hunk 6 above)
Comment (jmcatamney) [reply restore/restore.go]: Correct, in a successful restore everything should end on its own.  I can add an `if wasTerminated` check here.

File: end_to_end/end_to_end_suite_test.go (hunk 8 above)
Comment (jmcatamney) [reply end_to_end/end_to_end_suite_test.go]: Unless we want to get really paranoid and assert there are no `tail`s, no `gpbackup_helper`s, no pipes, and so forth, it's probably overkill.

File: backup/backup.go (hunk 10 above)
Comment (jmcatamney) [reply backup/backup.go]: Nope, good catch.

File: restore/remote.go (hunk 7 above)
Comment (jmcatamney) [reply restore/remote.go]: Yeah, the filename functions were in utils originally since backup and restore both use the same files, so if they're different we can move them to remote.go later.

File: restore/restore.go [hunk 11]
```diff
@@ -266,10 +268,12 @@ func DoCleanup() {
 	}()
 	logger.Verbose("Beginning cleanup")
 	if backupConfig.SingleDataFile {
-		CleanUpSegmentHelperProcesses(globalCluster)
 		CleanUpSegmentTOCs(globalCluster)
-		CleanUpHelperFilesOnAllHosts(globalCluster)
-		utils.TerminateHangingCopySessions(connection, globalCluster, "gprestore")
+		if wasTerminated { // These should all end on their own in a successful restore
+			CleanUpSegmentHelperProcesses(globalCluster)
+			CleanUpHelperFilesOnAllHosts(globalCluster)
```
Comment (chrishajas) [inline restore/restore.go]: We might always need to remove the oid and helper script files (outside the wasTerminated). I don't think they're removed somewhere else?

Comment (jmcatamney) [reply restore/restore.go]: Fixed.
